src/Npgsql/NpgsqlTypeLoadingOptions.cs

```diff
@@ -0,0 +1,57 @@
+namespace Npgsql;
+
+/// <summary>
+/// Options for configuring Npgsql type loading.
+/// </summary>
+public class NpgsqlTypeLoadingOptions
+{
+    readonly NpgsqlConnectionStringBuilder _csb;
+    internal NpgsqlTypeLoadingOptions(NpgsqlConnectionStringBuilder csb) => _csb = csb;
+
+    bool? _loadTableComposites;
+    /// <summary>
+    /// Load table composite type definitions, and not just free-standing composite types.
+    /// </summary>
+    public bool LoadTableComposites
```

Comment:
I think these should still be methods (e.g. EnableTableCompositeLoading()), so that they can return NpgsqlDataSourceBuilder - otherwise we break the fluent chaining:

```c#
npgsqlDataSourceBuilder
    .TypeLoading.EnableTableCompositeLoading(false)
    .UseNodaTime();
```

The alternative is to put the entire type loading in a lambda with its own builder:

```c#
npgsqlDataSourceBuilder
    .ConfigureTypeLoading(o => o.LoadTableComposite = false)
    .UseNodaTime();
```

The 2nd option still seems inconsistent - why does the outer builder have lambda-accepting methods with builders, but the inner builder (TypeLoadingOptionsBuilder?) have simple properties... In EF we've gone down the full builder everywhere route, which would be:

```c#
npgsqlDataSourceBuilder
    .ConfigureTypeLoading(o => o.EnableTableCompositeLoading(false))
    .UseNodaTime();
```

Whaddaya think?

Based on what we decide here I'll also do the tracing options...

Reply:
OK so I've changed it over to the builder pattern throughout, few things to mention:

1. `builder.ConnectionStringBuilder` is a big offender that won't fit in with the new configuration pattern.

2. For now I've implemented support for one configure action (called in `datasourcebuilder.Build()`). If we want to allow for type loading configuration coming from different places (e.g. our tests may want general setup and test specific setup) we have to change that to a list, or executing the actions immediately (the latter needs the unset property support again).

3. If we do support multiple configuration actions, do we really not want getters? Later actions might need to know what the current state is to configure things correctly. (even though it's not too relevant for this particular builder)

A broader issue, unrelated to one or the other configuration pattern is that we don't have access to the final DataSource configuration on the connection. Helpers like `IgnoreOnRedshift` (see changes) can't reliably pull relevant info from the connection anymore, as it may not be in the connection string. This is not a real issue yet, but we might end up bumping into this.

Reply:
> builder.ConnectionStringBuilder is a big offender that won't fit in with the new configuration pattern.

Indeed... Though since NpgsqlDataSourceBuilder accepts a connection string, it's maybe reasonable to expect people to prepare their connection string in advance and then instantiate their NpgsqlDataSourceBuilder with it. NpgsqlDataSourceBuilder.ConnectionStringBuilder is a bit of "just-in-case" thing IMHO, it's not great that it's not builder-based but maybe not terrible either.

> For now I've implemented support for one configure action [...]

On possible thought here is for the lambda to be executed immediately when ConfigureTypeLoading is executed, on an NpgsqlTypeLoadingOptions instance that NpgsqlDataSourceBuilder internally holds (rather than have it hold the lambda). This would allow executing ConfigureTypeLoading multiple times, with later invocations overriding earlier ones... Does that make sense?

> If we do support multiple configuration actions, do we really not want getters? Later actions might need to know what the current state is to configure things correctly. (even though it's not too relevant for this particular builder)

Good question.

More generally, I think we're moving in a direction where we'd eventually expose non-builder "options" types to the user (like NpgsqlTypeLoadingOptions); I think that's what we'd need to do to support Microsoft.Extensions.Options, for example. At that point, the builder itself could expose the current options object as well, so that users can read from it to know the current configuration state.

FWIW this is once again the EF strategy: you've got pure builders (which are intended to be the primary user-facing API), which perform their changes on lower-level "metadata" types which expose simple getters/setters (like our options types); the builder also exposes these metadata types via a Metadata property:

```c#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<Blog>().ToTable("foo");
    Console.WriteLine(modelBuilder.Entity<Blog>().Metadata.GetTableName());
}
```

> A broader issue, unrelated to one or the other configuration pattern is that we don't have access to the final DataSource configuration on the connection. Helpers like IgnoreOnRedshift (see changes) can't reliably pull relevant info from the connection anymore, as it may not be in the connection string. This is not a real issue yet, but we might end up bumping into this.

Right. So once we have proper, publicly-exposed options objects for everything (as above), we could also decide to expose read-only versions of these on NpgsqlDataSource (either internal just for our tests, or public)...